interpreter/mesh/src/main/scala/io/buoyant/interpreter/mesh/Client.scala

```diff
@@ -0,0 +1,324 @@
+    backoffs: scala.Stream[Duration],
+    timer: Timer
+  ): NameInterpreter with Delegator =
+    new Impl(root, interpreter, resolver, delegator, backoffs, timer)
+
+  private[this] val _releaseNop = () => Future.Unit
+  private[this] val _rescueUnit: PartialFunction[Throwable, Future[Unit]] = {
+    case NonFatal(_) => Future.Unit
+  }
+
+  private[this] class Impl(
+    root: Path,
+    interpreter: mesh.Interpreter,
+    resolver: mesh.Resolver,
+    delegator: mesh.Delegator,
+    backoffs: scala.Stream[Duration],
+    timer: Timer
+  ) extends NameInterpreter with Delegator {
+
+    /**
+     * When observed, streams bound trees from the mesh Interpreter.
+     *
+     * Leaf nodes include a Var[Addr] that, when observed, streams
+     * replica resolutions from the mesh Resolver.
+     */
+    override def bind(dtab: Dtab, path: Path): Activity[NameTree[Name.Bound]] = {
+      val open = () => interpreter.streamBoundTree(mkBindReq(root, path, dtab))
+      streamActivity(open, decodeBoundTree, backoffs, timer)
+    }
+
+    /**
+     * When observed, streams dtabs from the mesh server and
+     */
+    override lazy val dtab: Activity[Dtab] = {
+      val open = () => delegator.streamDtab(mkDtabReq(root))
+      streamActivity(open, decodeDtab, backoffs, timer)
+    }
+
+    override def delegate(
+      dtab: Dtab,
+      tree: NameTree[Name.Path]
+    ): Activity[DelegateTree[Name.Bound]] = {
+      val open = () => delegator.streamDelegateTree(mkDelegateTreeReq(root, dtab, tree))
+      streamActivity(open, decodeDelegateTree, backoffs, timer)
+    }
+
+    private[this] val resolve: Path => Var[Addr] = {
+      case Path.empty => Var.value(Addr.Failed("empty"))
+      case id =>
+        val open = () => resolver.streamReplicas(mkReplicasReq(id))
+        streamVar(Addr.Pending, open, replicasToAddr, backoffs, timer)
+    }
+
+    private[this] val fromBoundNameTree: mesh.BoundNameTree => NameTree[Name.Bound] =
+      mkFromBoundNameTree(resolve)
+
+    private[this] val decodeBoundTree: mesh.BoundTreeRsp => Option[NameTree[Name.Bound]] = {
+      case mesh.BoundTreeRsp(Some(ptree)) => Some(fromBoundNameTree(ptree))
+      case _ => None
+    }
+
+    private[this] val fromBoundDelegateTree: mesh.BoundDelegateTree => DelegateTree[Name.Bound] =
+      mkFromBoundDelegateTree(resolve)
+
+    private[this] val decodeDelegateTree: mesh.DelegateTreeRsp => Option[DelegateTree[Name.Bound]] = {
+      case mesh.DelegateTreeRsp(Some(ptree)) => Some(fromBoundDelegateTree(ptree))
+      case _ => None
+    }
+  }
+
+  /**
+   * Constructs a Var that, when observed, calls `open()` to obtain a
+   * gRPC stream, publishing updates into the Var's state.
+   *
+   * When the stream returns an error, the Var may be updated (as
+   * according to `toT`) and no further updates will will be
+   * published.
+   */
+  private[this] def streamVar[S, T](
+    init: T,
+    open: () => Stream[S],
+    toT: Try[S] => Option[T],
```

Comment:
We should either set the state on exceptions or change the signature of this to `S => Option[T]`